NatLibFi/ekirjasto-android-core
# Add test login page accessible using a deep link

The test login is disabled by default, and to enable it you need to add fields like these to `local.properties`:
```
ekirjasto.testLogin.enabled=true
ekirjasto.testLogin.username=TestUser
ekirjasto.testLogin.pinCode=1234567890
```

This enables the test login, which is then accessible via a link like:
`ekirjasto://test-login`

The username can be prefilled (the PIN code intentionally cannot):
`ekirjasto://test-login?username=TestUser`

The circulation API URL and library ID that should be used when the test login is active are defined in `gradle.properties` (can be overridden in `local.properties`), by default they're set to:
```
ekirjasto.testLogin.circulationApiUrl=https://lib-dev.e-kirjasto.fi
ekirjasto.testLogin.libraryProviderId=28bed937-a16b-4d69-a9c8-4b2656333423
```

This commit also changes:
- set `cleartextTrafficPermitted` to false, because allowing it gave a security warning in Google Play Console's prelaunch report
- quick fix for showing "null" as the version name on the settings page - the version name is hard-coded in the root-level build.gradle.kts, not sure if there's some better place to set it

Issue: [SIMPLYE-361](https://jira.lingsoft.fi/browse/SIMPLYE-361)


Testing
=======

Enable test login and add some test login credentials, both done in `local.properties`, for example:
```
ekirjasto.testLogin.enabled=true
ekirjasto.testLogin.username=AppReview
ekirjasto.testLogin.pinCode=1234567890
```

Build the app with some flavor other than the test login default of "dev", for example using "beta" (https://lib-beta.e-kirjasto.fi), which has production Suomi.fi login turned on as of writing.

Remember to wipe app data after changing flavors.

Open the app normally, check that Suomi.fi login doesn't have test accounts allowed. Login if you want, but the app caches the circulation URL immediately on open, so logging in is not necessary to test.

You can also confirm in the logs that the circulation URL is lib-beta.

Then, to test the test login:
- close the app
    - the app can't be open in the background, or the deep link will only bring the app to the foreground, instead of going to the test login page
- open the app using the test login deep link
    - I have a test link at https://wrbl.fi, so you can open that site in the Android device's web browser and click "E-kirjasto test link" from there

At the test login page, input the username and password that you defined in `local.properties` and press "Login".

The app will:
- delete all data directories except `shared_prefs`
    - this is where the app stores the info about whether the test login is active, so that can't be deleted
- restart the app

You should see the intro screen again. Check that Suomi.fi login now allows test accounts, and confirm in the logs that the circulation URL is now lib-dev instead of lib-beta.
